### PR TITLE
feat(api): add ability to create and authenticate users

### PR DESCRIPTION
**Pull request by preetpatel**
### Purpose
HeaRT needs the ability to create users as well as authenticate them.

[HRT-51](https://softeng761team5.atlassian.net/browse/HRT-51)
[HRT-55](https://softeng761team5.atlassian.net/browse/HRT-55)

### Approach
Create API endpoints to add a new user (/api/users) and an api to authenticate a user (/api/users/authenticate/:username) by POSTing with a json body containing "passwordHash", where the hash is computed using the bcrypt library on the frontend.

### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

#### Does this PR involve a UI change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: server/users/users.test.js [hunk 1]
```diff
@@ -21,16 +21,43 @@ describe("/api/users tests", () => {
...
-    ]);
+    const authResponse = await request(app)
+      .post("/api/users/authenticate/jdoe")
+      .send({
+        passwordHash: "aMoreSecureHashThanThisPlease",
+      });
+    expect(authResponse.status).toBe(200);
+    expect(authResponse.body.username).toEqual("jdoe");
   });
 });
```
Comment (dder725) [inline server/users/users.test.js:97]: Should there be a test for ensuring that an incorrect password does not result in successful auth?

**Review by dder725 — CHANGES_REQUESTED** [review]
Just one comment on testing, otherwise LGTM :)

File: server/users/users.test.js (hunk 1 above)
Comment (preetpatel) [reply server/users/users.test.js:97]: knew I was forgetting something! thanks

**Review by dder725 — DISMISSED** [review]

**Review by mitchfaulcon — CHANGES_REQUESTED** [review]
Should there also be a test for attempting to authorise a user that doesn't exist in the database? At the moment I never get a response from an incorrect username

**Review by mitchfaulcon — APPROVED** [review]
Thanks for that change, everything looks good now :)

**Review by dder725 — APPROVED** [review]

**Pull request merged**